esp-hal-common/src/interrupt/riscv.rs

```diff
@@ -600,6 +603,42 @@ unsafe fn get_assigned_cpu_interrupt(interrupt: Interrupt) -> CpuInterrupt {
 
     core::mem::transmute(cpu_intr)
 }
+#[cfg(feature = "interrupt-preemption")]
+#[inline(always)]
+unsafe fn handle_priority() -> u32 {
+    let interrupt_id: usize = mcause::read().code(); // MSB is whether its exception or interrupt.
+    #[cfg(not(esp32c6))]
+    let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+    #[cfg(esp32c6)]
+    let intr = &*crate::peripherals::INTPRI::PTR;
+    let interrupt_priority = intr
+        .cpu_int_pri_0
+        .as_ptr()
+        .offset(interrupt_id as isize)
+        .read_volatile();
+    let prev_interrupt_priority = intr.cpu_int_thresh.read().bits();
+    if interrupt_priority < 15 {
+        // leave interrupts disabled if interrupt is of max priority.
+        intr.cpu_int_thresh
+            .write(|w| w.bits(interrupt_priority + 1)); // set the prio threshold to 1 more than current interrupt prio
+        unsafe {
+            riscv::interrupt::enable();
+        }
+    }
+    prev_interrupt_priority
+}
+#[cfg(feature = "interrupt-preemption")]
+#[inline(always)]
+unsafe fn restore_priority(stored_prio: u32) {
+    #[cfg(not(esp32c6))]
+    let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+    #[cfg(esp32c6)]
+    let intr = &*crate::peripherals::INTPRI::PTR;
+    intr.cpu_int_thresh.write(|w| w.bits(stored_prio));
+    unsafe {
```

Comment:
I guess it would be safer if we disable interrupts before restoring the priority threshold

Reply:
Shouldn't it be fine to take an interrupt as long as we aren't in the middle of restoring context? I'll move it either way.